### PR TITLE
[engine] make authorizers field explicit in `Node.Exercise`

### PR DESCRIPTION
**Pull request by nickchapman-da**
Advances #15882

Add `choiceAuthorizers` field to `Node.Exercise`
```
      choiceAuthorizers: Option[Set[Party]],
```

This is a breaking change for canton.
The following branch contains the required changes:
https://github.com/DACH-NY/canton/compare/main...nick/adapt-to-authorizers-field-of-exe-node

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 1]
```diff
@@ -990,12 +990,8 @@ private[lf] object SBuiltin {
   /** $beginExercise
     *    :: arg                                           0 (choice argument)
     *    -> ContractId arg                                1 (contract to exercise)
-    *    -> List Party                                    2 (actors)
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:993]: Note: This change is just fixing a related comment which was just wrong/out-of-date.

**Comment by skisel-da** [comment]
In terms of migration of this new field on the exercise node - what is the value we get for the existing contracts? is it empty or should it be computed somehow?

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM

**Comment by nickchapman-da** [comment]
> In terms of migration of this new field on the exercise node - what is the value we get for the existing contracts? is it empty or should it be computed somehow?

YES. good question! I had been thinking about this... Semantically, the value of the `authorizers` field prior to it's existence was always: `contract.signatories union actingParties`. Now with an explicit field for `authorizers`, we can now represent any party-set, to indicate a restriction or gain in authority.

However, semantically, an empty set of `authorizers` should be rejected by the `AuthorizationChecker`; just as we do for `NoSignatories` and `NoControllers`.

This is very convenient because it allow us to give the representation `authorizers = Empty.set` an alternative meaning -- i.e. the old implicit default --  _contract.signatories union actingParties_ -- without it needing to be explicitly computed/represented in the node.

Taking this approach would also play nicely with `.proto` backwards compatibility. Assuming we represent `authorizers` with the `repeated` construct, then we can't distinguish _missing-field_ authorizers from _empty-set_ authorizers, anyway!

@andreaslochbihler-da Do you agree with this?

**Comment by andreaslochbihler-da** [comment]
> @andreaslochbihler-da Do you agree with this?

Yes, this makes a lot of sense and we avoid storing redundant information in the common case.

**Review by remyhaemmerle-da — APPROVED** [review]

**Comment by nickchapman-da** [comment]
On reflection, it will be cleaner to define the new `choiceAuthorizers` field as an `Option`
      choiceAuthorizers: Option[Set[Party]],

with `None` to indicating explicitly when the old/default behaviour occurs.

`Some(empty)` will never occur. And there Will be an authorization check for this.

We can still have the `.proto` representation discussed above.

**Comment by nickchapman-da** [comment]
> @nickchapman-da If this is really true, you could use the type `Option[NonEmpty[Set[Party]]]`.

Well... to be more precise... an empty set can occur, if the user's `authority` declaration evaluates to an empty set. But this will provoke a new authorization failure: `NoAuthorizers` . See: https://github.com/digital-asset/daml/pull/16658/files#diff-d2ac47f367673c1204cc28f7209d603f21eb1aa0ef0a78a8748e5dd6ff168b27R136

**Pull request merged**